instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter;
+
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.propagation.ContextPropagators;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import io.opentelemetry.instrumentation.api.InstrumentationVersion;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class InstrumenterBuilder<REQUEST, RESPONSE> {
+  private final OpenTelemetry openTelemetry;
+  private final String instrumentationName;
+  private InstrumenterConstructor<REQUEST, RESPONSE> constructor =
+      InstrumenterConstructor.internal();
+  private SpanNameExtractor<? super REQUEST> spanNameExtractor;
+  private SpanKind spanKind = SpanKind.INTERNAL;
+  private StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor =
+      StatusExtractor.getDefault();
+  private final List<AttributesExtractor<? super REQUEST, ? super RESPONSE>> attributesExtractors =
+      new ArrayList<>();
+  private ErrorCauseExtractor errorCauseExtractor = ErrorCauseExtractor.jdk();
+
+  InstrumenterBuilder(OpenTelemetry openTelemetry, String instrumentationName) {
+    this.openTelemetry = openTelemetry;
+    this.instrumentationName = instrumentationName;
+
+    // TODO: does this make any sense for a default? maybe it should be a mandatory param...
+    spanNameExtractor = request -> instrumentationName;
+  }
+
+  public InstrumenterBuilder<REQUEST, RESPONSE> setSpanNameExtractor(
+      SpanNameExtractor<? super REQUEST> spanNameExtractor) {
+    this.spanNameExtractor = spanNameExtractor;
+    return this;
+  }
+
+  public InstrumenterBuilder<REQUEST, RESPONSE> setClientSpanKind(TextMapSetter<REQUEST> setter) {
```

Comment:
I think we want something like `newClientInstrumenter` instead of these setters. Otherwise, calling this and the server span kind method together (which logically seems like it should work, they're different "things"), will cause the builder to break. So we want to end up with something like

```java
InstrumenterBuilder<> builder = builder
  .setSpanNameExtractor
  .setSpanStatusExtractor
  .addAttributesExtractor;

clientInstrumentor = builder.newClientInstrumenter(setter); 
// or builder.newClientInstrumenter(setter, new AwsSdkSpanKindExtractor())
serverInstrumenter = builder.newServerInstrumenter(getter);
```

I think it also removes the complicated InstrumenterConstructor class

Reply:
That being said I'm a bit worried having more differences in parameters - if `newClientInstrumenter` only ever needs 1-2 params, this pattern seems fine, if it needs 5, then we'd want a builder returned there... I think it's ok though can't expect this and we could do something in the future if running into such a situtaiton.

Reply:
Done 👍 

For the form that accepts a `SpanKindCreator` I'm thinking of avoiding the client/server words in the method name and instead calling them sth like:

```java
builder.newUpstreamInstrumenter(getter, spanKindExtractor);
builder.newDownstreamInstrumenter(setter, spanKindExtractor);
```

It kind of feels like calling it a "client instrumenter" when the type could be anything is a bit of a lie.